celery/canvas.py

```diff
@@ -1134,7 +1134,14 @@ def link_error(self, sig):
             # pass a Mock object as argument.
             sig['immutable'] = True
             sig = Signature.from_dict(sig)
-        return self.tasks[0].link_error(sig)
+        # Any child task might error so we need to ensure that they are all
+        # capable of calling the linked error signature. This opens the
+        # possibility that the task is called more than once but that's better
+        # than it not being called at all.
+        #
+        # We return a concretised tuple of the signatures actually applied to
+        # each child task signature, of which there might be none!
+        return tuple(child_task.link_error(sig) for child_task in self.tasks)
```

Comment:
If I understand correctly this means that the chord error handler may fire for every error in the chord header.  Is that accurate?

Reply:
This is just for groups, but if we substitute that in your comment then yep, that's exactly right. There's a little blurb in the commit and it is a potentially surprising change since previously a group errback would only be called if the 0th task failed. I toyed with the idea of adding a warning here for the changed behaviour but didn't end up doing it. That's certainly something which I could do if you all think that's worthwhile.

This comment made me have a look at what chords do for callbacks and errbacks, and basically they just get linked to the body task. Similar to group signatures, that's fine for callbacks since they only get called when everything swims along happily. The new integration tests do show that errbacks get called for failures in a chord header but they do only run with the Redis backend...

Reply:
Fantastic, thank you for that elaboration! 

The particular use case I had in mind mirrors the python `try: finally:` construct, something like this:

```
chord([might_fail.s(), might_fail.s(), will_finish.s(), ], finisher.s()).on_error(finisher.s())
```

That currently doesn't work (without the bug fix in this PR) because the chord error handler won't fire in case of error.  If the finisher would be called for every single error in the chord error, that wouldn't be much better, but as I understand it  should result in the finisher being called exactly once.

Reply:
Yep, that's my understanding as well. `test_errback_called_by_chord_from_group()` added in this PR validates that the errback is called at least once but doesn't count the number of calls for the two situations (errback on header group -> children, vs errback on the chord itself).

I think I should add a note to the docs about this, and perhaps modify the tests a bit to capture this behaviour. I've have a little think while I address the other comments here and work out if I want to do it on this PR.